vmm/src/vm.rs

```diff
@@ -968,6 +992,52 @@ impl Vm {
         Ok(EntryPoint { entry_addr })
     }
 
+    #[cfg(feature = "igvm")]
+    fn load_igvm(
+        igvm: File,
+        memory_manager: Arc<Mutex<MemoryManager>>,
+        cpu_manager: Arc<Mutex<cpu::CpuManager>>,
+    ) -> Result<EntryPoint> {
+        /*
+        BIOS-e820: [mem 0x0000000000000000-0x000000000009ffff] usable
+        BIOS-e820: [mem 0x00000000000a0000-0x00000000000fffff] reserved
+        BIOS-e820: [mem 0x0000000000100000-0x00000000001fffff] ACPI data
+        BIOS-e820: [mem 0x0000000001a00000-0x0000000003c79fff] usable
+         */
+        let cur_ram_size = memory_manager.lock().unwrap().current_ram();
+        debug!("current ram size: {:?}", cur_ram_size);
+        let mut arch_mem_regions: Vec<ArchMemRegion> = Vec::new();
+        arch_mem_regions.push(ArchMemRegion {
+            base: 0x00000000e8000000,
+            size: 0x10000000,
+            r_type: RegionType::Reserved,
+        });
+        if cur_ram_size <= 3 * GIGABYTE {
+            arch_mem_regions.push(ArchMemRegion {
+                base: 0x200000,
+                size: (cur_ram_size - 0x200000) as usize,
+                r_type: RegionType::Ram,
+            });
+        } else {
+            arch_mem_regions.push(ArchMemRegion {
+                base: 0x200000,
+                size: 3 * GIGABYTE as usize - 0x200000,
+                r_type: RegionType::Ram,
+            });
+            arch_mem_regions.push(ArchMemRegion {
+                base: 0x0000000100000000,
+                size: (cur_ram_size - 3 * GIGABYTE) as usize,
+                r_type: RegionType::Ram,
+            });
+        }
```

Comment:
Is this particular memory region layout required by the igvm? Don't you need to ensure it would be compatible with existing memory layout we are using from `arch::arch_memory_regions()` (which defines how we allocate the guest memory)?

Reply:
I would let @jinankjain  comment on this one.

Reply:
Umm it's complicated let me try to explain how memory map actually gets passed in case of IGVM and sev-snp enabled guest.

The [IGVM generator tool](https://github.com/microsoft/igvm-tooling) actually implements a bzImage/ELF loader and tries to load the Linux kernel in memory as per the Linux boot protocol. As part of this process, it also constructs the setup data which contains the information like e820 map which is the guest memory layout. But in case of sev-snp we do not want to describe the complete guest memory via that e820 map because it would extremely slow down the guest boot because all the ranges described in that e820 map needs to be pvalidated before the guest can actually use the memory. So, we tend to expose a very small early memory layout to the guest through which we only expose ranges which actually contains some meaningful data like kernel, initramfs, acpi tables etc. 

Now for the remaining guest memory ranges we have proposed a PV interface https://lkml.org/lkml/2022/11/9/1171 through which guest can extend its existing e820 map and pvalidate the remaining ranges after the critical boot setup is done. Since arch_mem_regions can be easily translated into what is required by this guest PV interface:

```
+struct memory_map_entry {
+	u64 starting_gpn;
+	u64 numpages;
+	u16 type;
+	u16 flags;
+	u32 reserved;
+};
```

Thus, we decided to overload the same interface to funnel that information. But if you think we are kind of abusing the existing interface we can define the new data structure which mimics this PV interface requirements and use that instead. But in general, I agree the ergonomics of this code can improved if we propose a new layout in arch/x86/layout.rs for IGVM and that would contain all these constants which describes the PCI hole and everything. We can do that as well.

Reply:
@jinankjain @likebreath Can we keep as is in this PR and improve it in follow up pr. I guess @jinankjain will take care of it.

Reply:
I don't think I am following the way how it works. If booting from a `IGVM` file requires a specific guest memory layout, how it would work without changing the way how the guest memory is allocated that now relies on `arch::arch_memory_regions()`?

Some basic questions: Does booting from `IGVM` files need firmware? Does `IGVM` file use its own boot protocol (e.g. something that is not PVH)? TBH, I am not familiar with boot process in general. /cc @rbradford is the expert in this domain.

With that, please go ahead and refactor this code for better readability. I am not sure if this piece of code should be moved to `arch/x86/layout.rs` or `vmm/src/igvm`. I guess it depends on how it will be used.

Reply:
@liuw for visibility

Reply:
@likebreath I can think of two options for now to move forward with this PR.
1. We can keep the mem layouts as is and start working on a better solution.
2. Keep the arch_mem_regions empty and pass it to the parser for now and start working on the solution.